read-fonts/src/tables/colr.rs

```diff
@@ -101,4 +157,686 @@ impl<'a> Colr<'a> {
+            for paint_record in base_glyph_records {
+                let gid = paint_record.glyph_id();
+                if !glyph_set.contains(GlyphId::from(gid)) {
+                    continue;
+                }
+                if let Ok(paint) = paint_record.paint(offset_data) {
+                    c.dispatch(&paint);
+                }
+            }
+
+            glyph_set.union(&c.glyph_set);
+        }
+
+        if let Some(Ok(clip_list)) = self.clip_list() {
+            c.glyph_set.union(glyph_set);
+            for clip_record in clip_list.clips() {
+                clip_record.v1_closure(&mut c, &clip_list);
+            }
+        }
+
+        //when a DeltaSetIndexMap is included, collected variation indices are actually delta set indices, we need to map them into variation indices
+        if let Some(Ok(var_index_map)) = self.var_index_map() {
+            delta_set_indices.extend(variation_indices.iter());
+            variation_indices.clear();
+            for idx in delta_set_indices.iter() {
+                if let Ok(var_idx) = var_index_map.get(idx) {
+                    let var_idx = ((var_idx.outer as u32) << 16) + var_idx.inner as u32;
+                    variation_indices.insert(var_idx);
+                }
+            }
+        }
+    }
+}
+
+struct Colrv1ClosureContext<'a> {
+    glyph_set: IntSet<GlyphId>,
+    layer_indices: &'a mut IntSet<u32>,
+    palette_indices: &'a mut IntSet<u16>,
+    variation_indices: &'a mut IntSet<u32>,
+    colr: &'a Colr<'a>,
+    nesting_level_left: u8,
+    visited_paints: IntSet<u32>,
+    colr_head: usize,
+}
+
+impl<'a> Colrv1ClosureContext<'a> {
+    pub fn new(
+        layer_indices: &'a mut IntSet<u32>,
+        palette_indices: &'a mut IntSet<u16>,
+        variation_indices: &'a mut IntSet<u32>,
+        colr: &'a Colr,
+    ) -> Self {
+        let colr_head = colr.offset_data().as_bytes().as_ptr() as usize;
+        Self {
+            glyph_set: IntSet::empty(),
+            layer_indices,
+            palette_indices,
+            variation_indices,
+            colr,
+            nesting_level_left: 64,
+            visited_paints: IntSet::empty(),
+            colr_head,
+        }
+    }
+
+    pub fn dispatch(&mut self, paint: &Paint) {
+        if self.nesting_level_left == 0 {
+            return;
+        }
+
+        if self.paint_visited(paint) {
+            return;
+        }
+        self.nesting_level_left -= 1;
+        paint.v1_closure(self);
+        self.nesting_level_left += 1;
+    }
+
+    fn paint_visited(&mut self, paint: &Paint) -> bool {
+        let offset = (paint.offset_data().as_bytes().as_ptr() as usize - self.colr_head) as u32;
+        if self.visited_paints.contains(offset) {
+            return true;
+        }
+
+        self.visited_paints.insert(offset);
+        false
+    }
+
+    pub fn add_layer_indices(&mut self, first_layer_index: u32, last_layer_index: u32) {
```

Comment:
does this need to be public API? Ditto for some of the other methods here, I think they're only called internally?